app/controllers/course/object_duplications_controller.rb

```diff
@@ -85,8 +85,13 @@ def load_videos_component_data
   end
 
   def load_destination_instances_data
-    @destination_instances = if can?(:duplicate_to_other_instances, current_tenant)
-                               Instance.all
+    @destination_instances = if can?(:duplicate_across_instances, current_tenant)
+                               instance_ids = InstanceUser.unscope(where: :instance_id).
```

Comment:
Why do you need to pluck the `instance_ids` first and query it again? 

Also, would not it be better if you first check if the current_user is an admin, then you can return the all instances directly. Otherwise, you are wasting the queries.

Also, take a look at CanCanCan's `accessible_by`. I think we can use that instead of repeating the logic from cancancan ability model here. You may need to re-define `duplicate_across_instances` similar to how `allow_group_manager_manage_group` is defined (instead of checking using blocks, you pass in the hash and let cancancan does its magic - let me know if you dont understand what I mean).

Reply:
> Why do you need to pluck the instance_ids first and query it again, instead of just doing something like Instance.where?

this can be done, however the code will become very ugly with all the `where` everywhere inside this one line, hence the splitting between `instance_ids` and, later on, fetching all `instances` having those `ids`

> Also, would not it be better if you first check if the current_user is an admin, then you can return the all instances directly. Otherwise, you are wasting the queries.

You're right on this. I have changed it accordingly. Though, performance-wise, it doesn't affect much since for each instance and user, only one instance_user exists there

> Also, take a look at CanCanCan's accessible_by. I think we can use that instead of repeating the logic from cancancan ability model here. You may need to re-define duplicate_across_instances similar to how allow_group_manager_manage_group is defined (instead of checking using blocks, you pass in the hash and let cancancan does its magic - let me know if you dont understand what I mean).

I have explored this options, however using hash for this is not a good idea as it makes the performance suffers (takes around 5 minutes just to generate all the permissions). This is because there's no limiting for the instance_users, and hence if using hash, instance will firstly attempt to load all the instance users and then pick those that satisfy the user_id as well as roles requirements.